### PR TITLE
♻️(frontend) use test wrappers in test

### LINKED ISSUE #2298: ♻️ Refactor frontend test suite
We recently add tools to simplify our frontend tests (#2264).

Now it's time to refactor our existing tests.

### PR DESCRIPTION
**Pull request by rlecellier**
We recently add tools to simplify our frontend tests (#2264). Now it's time to refactor our existing tests.

resolve #2298

DONE:
```
./widgets/UserLogin/components/UserMenu/index.spec.tsx
./widgets/UserLogin/index.not.isJoanieEnabled.spec.tsx
./widgets/UserLogin/index.spec.tsx
./widgets/Dashboard/hooks/useRouteInfo/index.spec.tsx
./widgets/Dashboard/hooks/useDashboardRouter/index.spec.tsx
./widgets/Dashboard/hooks/useEnroll/index.spec.tsx
./widgets/Dashboard/components/TeacherDashboardProfileSidebar/components/OrganizationLinks/index.spec.tsx
./widgets/Dashboard/components/TeacherDashboardProfileSidebar/index.spec.tsx
./widgets/Dashboard/components/NavigateWithParams/index.spec.tsx
./widgets/Dashboard/components/DashboardLayoutRoute/index.spec.tsx
./widgets/Dashboard/components/TeacherDashboardOrganizationSidebar/index.spec.tsx
./widgets/Dashboard/components/DashboardCard/index.spec.tsx
./widgets/Dashboard/components/DashboardItem/Contract/index.spec.tsx
./widgets/Dashboard/components/DashboardItem/Enrollment/DashboardItemEnrollment.spec.tsx
./widgets/Dashboard/components/DashboardItem/Enrollment/ProductCertificateFooter/index.spec.tsx
./widgets/Dashboard/components/DashboardItem/Certificate/index.spec.tsx
./widgets/Dashboard/components/DashboardItem/CourseEnrolling/index.spec.tsx
./widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.spec.tsx
./widgets/Dashboard/components/DashboardItem/Order/OrderStateTeacherMessage/index.spec.tsx
./widgets/Dashboard/components/DashboardItem/Order/OrderStateLearnerMessage/index.spec.tsx
./widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrderContract.useUnionResource.cache.spec.tsx
./widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrderContract.spec.tsx
./widgets/Dashboard/components/DashboardItem/CertificateStatus/index.spec.tsx
./widgets/Dashboard/components/DashboardItem/index.spec.tsx
./widgets/Dashboard/components/TeacherDashboardCourseSidebar/index.spec.tsx
./widgets/Dashboard/components/RouterButton/index.spec.tsx
./widgets/Dashboard/components/DashboardSidebar/components/MenuNavLink/index.spec.tsx
./widgets/Dashboard/components/DashboardSidebar/components/ContractNavLink/index.spec.tsx
./widgets/Dashboard/components/DashboardBox/index.spec.tsx
./widgets/Dashboard/components/DashboardAvatar/index.spec.tsx
./widgets/Dashboard/components/ProtectedOutlet/ProtectedOutlet.spec.tsx
./widgets/Dashboard/components/ProtectedOutlet/AuthenticatedOutlet.spec.tsx
./widgets/Dashboard/index.spec.tsx
./widgets/Search/hooks/useCourseSearch/index.spec.tsx
./widgets/Search/hooks/useFilterValue/index.spec.tsx
./widgets/Search/components/SearchFiltersPane/index.spec.tsx
./widgets/Search/components/SearchFilterGroupModal/index.spec.tsx
./widgets/Search/components/SearchFilterValueParent/index.spec.tsx
./widgets/Search/components/PaginateCourseSearch/index.spec.tsx
./widgets/Search/components/SearchFilterGroup/index.spec.tsx
./widgets/Search/components/SearchFilterValueLeaf/index.spec.tsx
./widgets/Search/index.spec.tsx
./widgets/SearchSuggestField/index.spec.tsx
./widgets/RootSearchSuggestField/index.spec.tsx
./widgets/SyllabusCourseRunsList/hooks/useCourseEnrollment/index.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseWishButton/index.login.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseWishButton/hooks/useCourseWish/index.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseWishButton/index.logout.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseRunEnrollment/CourseRunUnenrollmentButton/index.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseRunEnrollment/index.joanie.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseRunEnrollment/index.openedx.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseProductItem/components/CourseProductCourseRuns/index.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseProductItem/components/CourseProductCertificateItem/index.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseProductItem/components/CourseRunItem/index.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseProductItem/index.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseRunItem/index.spec.tsx
./widgets/SyllabusCourseRunsList/components/CourseRunItemWithEnrollment/index.spec.tsx
./widgets/SyllabusCourseRunsList/components/SyllabusSimpleCourseRunsList/index.spec.tsx
./widgets/SyllabusCourseRunsList/index.spec.tsx
./widgets/LanguageSelector/index.spec.tsx
./widgets/LtiConsumer/index.spec.tsx
 ```
 
TODO:
```
./widgets/index.spec.tsx
./contexts/JoanieApiContext/index.spec.tsx
./contexts/SessionContext/no-authentication.spec.tsx
./contexts/SessionContext/JoanieSessionProvider.spec.tsx
./contexts/SessionContext/index.spec.tsx
./hooks/useCourseSearchParams/index.spec.tsx
./hooks/useCourseProductUnion/index.spec.tsx
./hooks/useDefaultOrganizationId/index.spec.tsx
./hooks/useCreditCards/index.spec.tsx
./hooks/useDownloadCertificate/index.spec.tsx
./hooks/useProductOrder/index.spec.tsx
./hooks/useContractArchive/index.download.spec.tsx
./hooks/useContractArchive/index.spec.tsx
./hooks/useJoanieUserAbilities/index.not.isJoanieEnabled.spec.tsx
./hooks/useJoanieUserAbilities/index.spec.tsx
./hooks/useStaticFilters/index.spec.tsx
./hooks/useUnionResource/index.spec.tsx
./hooks/useHistory/index.spec.tsx
./hooks/useCourses/index.spec.tsx
./hooks/useResources/index.spec.tsx
./pages/DashboardAddressesManagement/DashboardCreateAddress.spec.tsx
./pages/DashboardAddressesManagement/DashboardEditAddress.spec.tsx
./pages/DashboardAddressesManagement/index.spec.tsx
./pages/TeacherDashboardCourseLearnersLayout/hooks/useCourseLearnersFilters/index.spec.tsx
./pages/TeacherDashboardCourseLearnersLayout/components/CourseLearnersFiltersBar/index.spec.tsx
./pages/TeacherDashboardCourseLearnersLayout/components/CourseLearnerDataGrid/index.spec.tsx
./pages/TeacherDashboardCourseLearnersLayout/index.spec.tsx
./pages/TeacherDashboardTraining/index.spec.tsx
./pages/TeacherDashboardCourseLoader/CourseRunList/CourseRunListCell/index.spec.tsx
./pages/TeacherDashboardCourseLoader/CourseRunList/utils.spec.tsx
./pages/TeacherDashboardCourseLoader/CourseRunList/index.spec.tsx
./pages/DashboardOrderLayout/index.spec.tsx
./pages/DashboardContracts/index.spec.tsx
./pages/DashboardCourses/index.spec.tsx
./pages/DashboardCertificates/index.spec.tsx
./pages/TeacherDashboardContractsLayout/hooks/useHasContractToDownload/index.spec.tsx
./pages/TeacherDashboardContractsLayout/hooks/useDownloadContractArchive/index.spec.tsx
./pages/TeacherDashboardContractsLayout/hooks/useTeacherContractFilters/index.spec.tsx
./pages/TeacherDashboardContractsLayout/hooks/useCheckContractArchiveExists/index.spec.tsx
./pages/TeacherDashboardContractsLayout/TeacherDashboardContracts/index.spec.tsx
./pages/TeacherDashboardContractsLayout/components/SignOrganizationContractButton/index.spec.tsx
./pages/TeacherDashboardContractsLayout/components/ContractFiltersBar/index.spec.tsx
./pages/TeacherDashboardContractsLayout/components/BulkDownloadContractButton/index.timer.spec.tsx
./pages/TeacherDashboardContractsLayout/components/BulkDownloadContractButton/index.spec.tsx
./pages/TeacherDashboardContractsLayout/components/ContractActionsBar/index.spec.tsx
./pages/TeacherDashboardCoursesLoader/index.spec.tsx
./pages/DashboardCreditCardsManagement/DashboardEditCreditCard.spec.tsx
./pages/DashboardCreditCardsManagement/index.spec.tsx
./utils/CreditCardHelper/index.spec.tsx
./utils/JoinAnd/index.spec.tsx
./utils/react-query/useSessionQuery/index.spec.tsx
./utils/react-query/useSessionMutation/index.spec.tsx
./utils/IntlHelper/index.spec.tsx
./utils/StringHelper/index.spec.tsx
./utils/CourseRuns/index.spec.tsx
./components/Modal/index.spec.tsx
./components/DjangoCMSTemplate/index.spec.tsx
./components/PaymentButton/components/PaymentInterfaces/index.spec.tsx
./components/PaymentButton/index.spec.tsx
./components/Badge/index.spec.tsx
./components/AddressesManagement/AddressForm/index.spec.tsx
./components/AddressesManagement/index.spec.tsx
./components/PurchaseButton/index.spec.tsx
./components/SaleTunnel/components/SaleTunnelStepPayment/index.spec.tsx
./components/SaleTunnel/components/SaleTunnelStepResume/index.spec.tsx
./components/SaleTunnel/components/RegisteredCreditCard/index.spec.tsx
./components/SaleTunnel/components/SaleTunnelStepValidation/index.spec.tsx
./components/SaleTunnel/components/StepBreadcrumb/index.spec.tsx
./components/SaleTunnel/index.spec.tsx
./components/DownloadContractButton/index.spec.tsx
./components/CourseGlimpse/index.spec.tsx
./components/CourseGlimpseList/index.spec.tsx
./components/Banner/index.spec.tsx
./components/Form/Select/index.spec.tsx
./components/Form/Input/index.spec.tsx
./components/ContractFrame/OrganizationContractFrame.spec.tsx
./components/ContractFrame/AbstractContractFrame.spec.tsx
./components/ContractFrame/LearnerContractFrame.spec.tsx
./components/Icon/index.spec.tsx
./components/SignContractButton/index.omniscientOrders.spec.tsx
./components/SignContractButton/index.spec.tsx
./components/ContractStatus/index.spec.tsx
./components/SearchInput/index.spec.tsx
./components/Pagination/index.spec.tsx
./components/TeacherDashboardCourseList/index.spec.tsx
./components/RegisteredAddress/index.spec.tsx
```

### CODE REVIEW

File: src/frontend/js/utils/test/wrappers/ReactQueryAppWrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  children,
+  intlOptions,
+  queryOptions,
+}: PropsWithChildren<AppWrapperProps>) => {
+  return (
+    <IntlWrapper {...(intlOptions || { locale: 'en' })}>
+      <ReactQueryWrapper {...(queryOptions || {})}>{children}</ReactQueryWrapper>
+    </IntlWrapper>
+  );
+};
```
Comment (jbpenrath) [inline src/frontend/js/utils/test/wrappers/ReactQueryAppWrapper.tsx:16]: Just a suggestion to improve readability.

```suggestion
export const ReactQueryAppWrapper = ({
  children,
  intlOptions = { locale: 'en' },
  queryOptions = {},
}: PropsWithChildren<AppWrapperProps>) => {
  return (
    <IntlWrapper {...intlOptions}>
      <ReactQueryWrapper {...queryOptions}>{children}</ReactQueryWrapper>
    </IntlWrapper>
  );
};
```

File: src/frontend/js/utils/test/render.tsx [hunk 2]
```diff
@@ -65,7 +65,11 @@ export const render: RenderFunction = (
       rerenderOptions?: Partial<Omit<AppWrapperProps, 'testingLibraryOptions'>>,
     ) => {
       return renderResult.rerender(
-        Wrapper === null ? element : <Wrapper {...rerenderOptions}>{rerenderElement}</Wrapper>,
+        Wrapper === null ? (
+          rerenderElement
+        ) : (
+          <Wrapper {...rerenderOptions}>{rerenderElement}</Wrapper>
+        ),
```
Comment (jbpenrath) [inline src/frontend/js/utils/test/render.tsx:72]: But as the first render renders the element with Wrapper, the rerender should do the same thing nope?

File: src/frontend/js/widgets/Dashboard/components/DashboardAvatar/index.spec.tsx [hunk 3]
```diff
@@ -1,15 +1,16 @@
-import { render, screen } from '@testing-library/react';
+import { screen } from '@testing-library/react';
 import { JoanieFileFactory } from 'utils/test/factories/joanie';
+import { render } from 'utils/test/render';
 import { DashboardAvatar, DashboardAvatarVariantEnum } from '.';
 
 describe('<DashboardAvatar/>', () => {
   it('should work with empty title', () => {
-    render(<DashboardAvatar title="" />);
+    render(<DashboardAvatar title="" />, { wrapper: null });
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardAvatar/index.spec.tsx:8]: Maybe we could make options optional ?

Comment (rlecellier) [reply src/frontend/js/widgets/Dashboard/components/DashboardAvatar/index.spec.tsx:8]: `options` is optional.
By default it's :
```ts
const toto = {
  wrapper: JoanieAppWrapper,
  queryOptions: {
    client: createTestQueryClient({ user: true })
  }
  intlOption: {
    locale: 'en',
  }
}
```

File: src/frontend/js/utils/test/render.tsx (hunk 2 above)
Comment (rlecellier) [reply src/frontend/js/utils/test/render.tsx:72]: it's already the case, if it was render with wrapper the first time, it'll be rerender with wrapper.

File: src/frontend/js/widgets/Dashboard/components/DashboardAvatar/index.spec.tsx (hunk 3 above)
Comment (jbpenrath) [reply src/frontend/js/widgets/Dashboard/components/DashboardAvatar/index.spec.tsx:8]: Oh right understood

File: src/frontend/js/utils/test/render.tsx (hunk 2 above)
Comment (jbpenrath) [reply src/frontend/js/utils/test/render.tsx:72]: I just don't understand why you need to wrap again the element in the wrapper 🤔

Comment (rlecellier) [reply src/frontend/js/utils/test/render.tsx:72]: it's how rerender works. This way it's allow us to update props on rerender.

```ts
(property) rerender: (ui: React.ReactNode) => void
```

**Review by jbpenrath — APPROVED** [review]
LGTM ✅

**Pull request merged**